ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/StreamStreamJoin.java

```diff
@@ -14,45 +14,86 @@
 
 package io.confluent.ksql.execution.plan;
 
+import static java.util.Objects.requireNonNull;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.common.collect.ImmutableList;
 import com.google.errorprone.annotations.Immutable;
+import io.confluent.ksql.name.ColumnName;
 import java.time.Duration;
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 
 @Immutable
 public class StreamStreamJoin<K> implements ExecutionStep<KStreamHolder<K>> {
 
+  // keyName was not present before 0.10.0, defaults to legacy ROWKEY
+  // This can be removed with the next breaking change.
+  public static final String LEGACY_KEY_COL = "ROWKEY";
+
   private final ExecutionStepPropertiesV1 properties;
   private final JoinType joinType;
+  private final ColumnName keyColName;
   private final Formats leftInternalFormats;
   private final Formats rightInternalFormats;
   private final ExecutionStep<KStreamHolder<K>> leftSource;
   private final ExecutionStep<KStreamHolder<K>> rightSource;
   private final Duration beforeMillis;
   private final Duration afterMillis;
 
-  public StreamStreamJoin(
+  @SuppressWarnings("unused") // Invoked by reflection
+  @JsonCreator
+  @Deprecated() // Can be removed at next incompatible version
+  private StreamStreamJoin(
```

Comment:
can jackson discover private constructors? clearly the tests pass, but I'm surprised

Reply:
Yeah, Jackson can access private constructors, methods and even fields directly.